### PR TITLE
Oauth2

### PR DESCRIPTION
**Pull request by freezey**
Add ability to fallback from OAuth2 connection to API Key connection automatically.
Tweak the TokenStorageProvider interaction so that the entire OAuth process, including refreshing tokens, can happen outside of the Infusionsoft SDK via the TokenStorageProvider class that is passed into the App

### CODE REVIEW

File: Infusionsoft/App.php [hunk 1]
```diff
@@ -196,8 +230,16 @@ public function hasTokens(){
     }
 
     public function refreshTokens(){
-        $tokens = Infusionsoft_OAuth2::refreshToken($this->refreshToken);
-        $this->updateAndSaveTokens($tokens['access_token'], $tokens['refresh_token'], $tokens['expires_in']);
+        if (method_exists($this->tokenStorageProvider, 'refreshTokens')){
```
Comment (joeynovak) [inline Infusionsoft/App.php]: This seems fairly hokey.  A token storage provider shouldn't refresh the tokens.

File: Infusionsoft/App.php [hunk 2]
```diff
@@ -196,8 +230,16 @@ public function hasTokens(){
     }
 
     public function refreshTokens(){
-        $tokens = Infusionsoft_OAuth2::refreshToken($this->refreshToken);
-        $this->updateAndSaveTokens($tokens['access_token'], $tokens['refresh_token'], $tokens['expires_in']);
+        if (method_exists($this->tokenStorageProvider, 'refreshTokens')){
+            $tokens = $this->tokenStorageProvider->refreshTokens();
+            $this->accessToken = $tokens['accessToken'];
```
Comment (joeynovak) [inline Infusionsoft/App.php]: This is also kind of weird.  I think it should return true / false, and then you call get tokens to get the updated tokens.  (Yes, you did follow the way it was written, it's wrong too IMO)

File: Infusionsoft/TokenStorageProvider.php [hunk 3]
```diff
@@ -9,5 +9,4 @@
 interface Infusionsoft_TokenStorageProvider {
     public function saveTokens($appDomainName, $accessToken, $refreshToken, $expiresIn);
     public function getTokens($appDomainName);
-    public function getFirstAppName();
```
Comment (joeynovak) [inline Infusionsoft/TokenStorageProvider.php]: Why did this get deleted?

Comment (freezey) [inline Infusionsoft/TokenStorageProvider.php]: Our implementation of the TokenStorageProvider only allows the correct app to authenticate, we never reference getFirstAppName and it did not seem appropriate to require it in TokenStorageProvider classes anymore

File: Infusionsoft/App.php (hunk 1 above)
Comment (freezey) [inline Infusionsoft/App.php]: Hmm - The alternative that I can see is having a TokenHandler class designation that supersedes using the TokenStorageProvider classes. A TokenHandler class would over-ride all OAuth2 interaction from the SDK (as we are doing with the current implementation of NovakSolutionsInfusionsoftTokenStorageProvider) while TokenStorageProvider would do no more than store and retrieve tokens.  Thoughts?

**Pull request merged**